src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2236 @@
+  exprt starts_with_digit=and_exprt
+    (binary_relation_exprt(chr, ID_ge, zero_char),
+     binary_relation_exprt(chr, ID_le, nine_char));
+  or_exprt a2(starts_with_digit, starts_with_minus);
+  axioms.push_back(a2);
+
+  for(size_t size=1; size<=max_size; size++)
+  {
+    // For each possible size, we add axioms:
+    // all_numbers: forall 1<=i<=size. '0'<=res[i]<='9'
+    // a3 : |res|=size&&'0'<=res[0]<='9' =>
+    //      i=sum+str[0]-'0' && all_numbers
+    // a4 : |res|=size&&res[0]='-' => i=-sum
+    // a5 : size>1 => |res|=size&&'0'<=res[0]<='9' => res[0]!='0'
+    // a6 : size>1 => |res|=size&&res[0]'-' => res[1]!='0'
+    // a7 : size==max_size => i>1000000000
+    exprt sum=from_integer(0, type);
+    exprt all_numbers=true_exprt();
+    chr=res[0];
+    exprt first_value=typecast_exprt(minus_exprt(chr, zero_char), type);
+
+    for(size_t j=1; j<size; j++)
+    {
+      chr=res[j];
+      sum=plus_exprt(
+        mult_exprt(sum, ten),
+        typecast_exprt(minus_exprt(chr, zero_char), type));
+      first_value=mult_exprt(first_value, ten);
+      and_exprt is_number(
+        binary_relation_exprt(chr, ID_ge, zero_char),
+        binary_relation_exprt(chr, ID_le, nine_char));
+      all_numbers=and_exprt(all_numbers, is_number);
+    }
+
+    axioms.push_back(all_numbers);
+
+    equal_exprt premise=res.axiom_for_has_length(size);
+    equal_exprt constr3(i, plus_exprt(sum, first_value));
+    implies_exprt a3(and_exprt(premise, starts_with_digit), constr3);
+    axioms.push_back(a3);
+
+    implies_exprt a4(
+      and_exprt(premise, starts_with_minus),
+      equal_exprt(i, unary_minus_exprt(sum)));
+    axioms.push_back(a4);
+
+    // disallow 0s at the beginning
+    if(size>1)
+    {
+      equal_exprt r0_zero(res[zero], zero_char);
+      implies_exprt a5(
+        and_exprt(premise, starts_with_digit),
+        not_exprt(r0_zero));
+      axioms.push_back(a5);
+
+      exprt one=refined_string_typet::index_of_int(1);
+      equal_exprt r1_zero(res[one], zero_char);
+      implies_exprt a6(
+        and_exprt(premise, starts_with_minus),
+        not_exprt(r1_zero));
+      axioms.push_back(a6);
+    }
+
+    // we have to be careful when exceeding the maximal size of integers
+    // Warning this should be different depending on max size
+    if(size==max_size)
+    {
+      exprt smallest_with_10_digits=from_integer(
+        smallest_by_digit(max_size), type);
+      binary_relation_exprt big(i, ID_ge, smallest_with_10_digits);
+      implies_exprt a7(premise, big);
+      axioms.push_back(a7);
+    }
+  }
+  return res;
+}
+
+
+exprt string_constraint_generatort::int_of_hex_char(
+  exprt chr, unsigned char_width, typet char_type) const
+{
+  exprt zero_char=constant_char('0');
+  exprt nine_char=constant_char('9');
+  exprt a_char=constant_char('a');
+  return if_exprt(
+    binary_relation_exprt(chr, ID_gt, nine_char),
+    plus_exprt(constant_char(10), minus_exprt(chr, a_char)),
+    minus_exprt(chr, zero_char));
```

Comment:
Actually this is used for the conversion from int to string and java only use characters 0123456789abcdef https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#toHexString(int) 
I've added a comment header which makes this explicit:
https://github.com/diffblue/cbmc/pull/374/commits/f4de9309326b308123eb2a34ae52ff7daa0d82df